File: go/internal/metrics/metrics.go
package metrics

import (
	"context"
	"fmt"
	"runtime"
	"slices"
	"strconv"
	"sync"

	"github.com/redis/go-redis/v9"
)

type Type string

const (
	MemoryStart   Type = "start"
	MemoryEnd     Type = "end"
	MemoryProcess Type = "process"

	Limit = 500
)

type MetricsUnit struct {
	Data  []uint64
	Mutex sync.RWMutex
}

type Metrics struct {
	redis *redis.Client
}

func New(redis *redis.Client) *Metrics {
	return &Metrics{
		redis: redis,
	}
}

func (m *Metrics) Save(ctx context.Context, unitType Type, value uint64) {
	go func() {
		m.redis.RPush(ctx, normalizeKey(unitType), value)
	}()
}

func (m *Metrics) GetStats(ctx context.Context, unitType Type) map[string]any {
	unit, err := m.redis.LRange(ctx, normalizeKey(unitType), 0, -1).Result()
	if err != nil || len(unit) == 0 {
		return map[string]any{
			"raw": map[string]uint64{
				"max": 0,
				"min": 0,
				"avg": 0,
				"p95": 0,
			},
			"formatted": map[string]string{
				"max": formatBytes(0, 2),
				"min": formatBytes(0, 2),
				"avg": formatBytes(0, 2),
				"p95": formatBytes(0, 2),
			},
			"values": []uint64{},
			"count":  0,
		}
	}

	sorted := make([]uint64, len(unit))
	for i, v := range unit {
		sorted[i], _ = strconv.ParseUint(v, 10, 64)
	}
	slices.Sort(sorted)

	count := len(sorted)
	max := sorted[count-1]
	min := sorted[0]

	var sum uint64
	for _, v := range sorted {
		sum += v
	}
	avg := sum / uint64(count)

	p95Index := int(float64(count) * 0.95)
	if p95Index >= count {
		p95Index = count - 1
	}
	p95 := sorted[p95Index]

	lastValues := sorted
	if len(sorted) > 50 {
		lastValues = sorted[len(sorted)-50:]
	}

	return map[string]any{
		"raw": map[string]uint64{
			"max": max,
			"min": min,
			"avg": avg,
			"p95": p95,
		},
		"formatted": map[string]string{
			"max": formatBytes(max, 2),
			"min": formatBytes(min, 2),
			"avg": formatBytes(avg, 2),
			"p95": formatBytes(p95, 2),
		},
		"values": lastValues,
		"count":  count,
	}
}

func GetMemoryUsage() uint64 {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	return m.Alloc
}

func formatBytes(bytes uint64, precision int) string {
	units := []string{"B", "KB", "MB", "GB", "TB"}

	value := float64(bytes)
	i := 0
	for value > 1024 && i < len(units)-1 {
		value /= 1024
		i++
	}

	return fmt.Sprintf("%.*f %s", precision, value, units[i])
}

func normalizeKey(key Type) string {
	return "go:memory:" + string(key)
}

File: node/lib/metrics.js
const redis = require('./redis');

const METRIC_PREFIX = 'node:memory:';
const MetricsType = {
  MemoryStart: 'start',
  MemoryProcess: 'process',
  MemoryEnd: 'end',
};

function getMemoryUsage() {
  return process.memoryUsage().heapUsed;
}

function formatBytes(bytes, precision = 2) {
  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  let value = bytes;
  let i = 0;

  while (value > 1024 && i < units.length - 1) {
    value /= 1024;
    i++;
  }

  return `${value.toFixed(precision)} ${units[i]}`;
}

async function saveMetric(type, value) {
  try {
    await redis.rPush(METRIC_PREFIX + type, String(value));
  } catch (e) {
    console.error('Failed to save metric', type, e.message);
  }
}

async function getStats(type) {
  try {
    const values = await redis.lRange(METRIC_PREFIX + type, 0, -1);

    if (!values || values.length === 0) {
      return {
        raw: { max: 0, min: 0, avg: 0, p95: 0 },
        formatted: { max: formatBytes(0), min: formatBytes(0), avg: formatBytes(0), p95: formatBytes(0) },
        values: [],
        count: 0
      };
    }

    const sorted = values.map(v => parseInt(v, 10)).sort((a, b) => a - b);
    const count = sorted.length;

    if (count === 0) {
      return {
        raw: { max: 0, min: 0, avg: 0, p95: 0 },
        formatted: { max: formatBytes(0), min: formatBytes(0), avg: formatBytes(0), p95: formatBytes(0) },
        values: [],
        count: 0
      };
    }

    const max = sorted[count - 1];
    const min = sorted[0];
    const sum = sorted.reduce((acc, val) => acc + val, 0);
    const avg = Math.floor(sum / count);

    const p95Index = Math.floor(count * 0.95);
    const p95 = sorted[Math.min(p95Index, count - 1)];

    const lastValues = sorted.length > 50 ? sorted.slice(-50) : sorted;

    return {
      raw: { max, min, avg, p95 },
      formatted: {
        max: formatBytes(max),
        min: formatBytes(min),
        avg: formatBytes(avg),
        p95: formatBytes(p95)
      },
      values: lastValues,
      count
    };
  } catch (e) {
    console.error('Failed to get stats:', e.message);
    return {
      raw: { max: 0, min: 0, avg: 0, p95: 0 },
      formatted: { max: formatBytes(0), min: formatBytes(0), avg: formatBytes(0), p95: formatBytes(0) },
      values: [],
      count: 0
    };
  }
}

module.exports = {
  MetricsType,
  getMemoryUsage,
  saveMetric,
  getStats
};
